### PR TITLE
intersphinx: Reduce log message severity when ambiguous target resolves to a duplicate

### LINKED ISSUE #12585: Warning about duplicate definitions in intersphinx mapping
### Describe the bug

We recently started getting warnings about duplicate definition in intersphinx, from two upstream projects:

- sklearn, see https://github.com/scikit-learn/scikit-learn/issues/29337
- ipywidgets, see https://github.com/jupyter-widgets/ipywidgets/issues/3930

Sklearn maintainers suspect this to be a sphinx bug, so I am opening this issue here.

### How to Reproduce
```
$ pip list | grep  Sphinx
Sphinx                        7.4.2
$ python -m sphinx.ext.intersphinx https://scikit-learn.org/stable/objects.inv | grep "    y "
WARNING:sphinx.sphinx.util.inventory:inventory <> contains multiple definitions for std:term:y
    y        
```

and
```
$ python -m sphinx.ext.intersphinx https://ipywidgets.readthedocs.io/en/stable/objects.inv | grep 'Widget Layout.ipynb#display'
WARNING:sphinx.sphinx.util.inventory:inventory <> contains multiple definitions for std:label:examples/Widget Layout.ipynb#display
    examples/Widget Layout.ipynb#display     display                                 : examples/Widget%20Layout.html#id1
```

### Environment Information

```text
Sphinx 7.4.2
```

### Sphinx extensions

```python
intersphinx
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jayaddison**
### Feature or Bugfix
- Bugfix

### Purpose
- Don't emit warnings about ambiguous entries that are pure-duplicates during resolution of Intersphinx references.

### Detail
- ~~Use `set`-comparison of the reference fields.~~ however, do use a set to count the number of distinct matched definitions at resolution-time. 
- Do log a `debug` messages for reference.

### Relates
- Warning introduced in #12329.
- Extends from #12586.

Edit: redraft description since #12586 was merged separately.
Edit: remove resolves-issue reference.

### CODE REVIEW

**Comment by jayaddison** [comment]
@AA-Turner a note to let you know that I'm holding this change in reserve - it is about Intersphinx _resolution_ -- e.g. occurs when people attempt to include links to potential ambiguity in external projects from their own.

This should happen far less often than ambiguity-detection at inventory loading-time, and I think is a more serious scenario (because it could result in incorrect hyperlinks).  Even so, the pure-duplicate case might make sense to reduce log-severity for here too.

**Comment by jayaddison** [comment]
(I'm not sure that I explained that particularly clearly, so let me know if I can clarify.  there are a few similar-seeming changes in-flight here, and I regret that because I could have planned this rollout better to reduce the chance of confusion)

**Comment by jayaddison** [comment]
Following the principle of "who is this warning actionable for" (ref [here](https://github.com/sphinx-doc/sphinx/issues/12613#issuecomment-2236846759)), I don't think that either the referrer or the referee are likely to want or need to do anything about ambiguities that are effectively duplicate entries.  It's potentially useful to be aware of these cases, but not something that should block strict-mode builds.

**Comment by jayaddison** [comment]
NB: This has no merge conflicts, but v7.4.6 has already been released, so the `CHANGES.rst` entry's placement is incorrect.

**Comment by jayaddison** [comment]
> NB: This has no merge conflicts, but v7.4.6 has already been released, so the `CHANGES.rst` entry's placement is incorrect.

(this is now resolved, so I've removed the do-not-merge label from this pull request)

**Comment by AA-Turner** [comment]
Does this address a regression in 7.4? Otherwise I will hold off until 7.5 // 8.0.

A

**Comment by jayaddison** [comment]
> Does this address a regression in 7.4? Otherwise I will hold off until 7.5 // 8.0.

It does not, no - and that sounds fine to me, thank you :+1:

**Review by firefly9966 — APPROVED** [review]

**Comment by jayaddison** [comment]
Thanks again!
